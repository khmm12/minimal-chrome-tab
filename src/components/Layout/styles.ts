import * as p from 'styled-system/patterns'

export const container = p.grid.raw({
  height: 'full',
  width: 'full',
  gridTemplateRows: '1fr auto 1fr',
  gap: '4rem',
  padding: '4rem',
  lineHeight: 0.8,
  fontFamily: 'Digital7Mono',
  color: { base: 'black', _dark: 'white' },
})
