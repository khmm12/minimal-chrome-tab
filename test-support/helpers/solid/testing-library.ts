import type { JSX } from 'solid-js'
import { hydrate as solidHydrate, render as solidRender } from 'solid-js/web'
import { getQueriesForElement, prettyDOM, queries as allQueries } from '@testing-library/dom'
import type { Queries, BoundFunction, prettyFormat } from '@testing-library/dom'

interface Ref {
  container: HTMLElement
  unmount: () => void
  dispose: () => void
}

type Ui = () => JSX.Element

interface Options<Q extends Queries = typeof allQueries> {
  container?: HTMLElement
  baseElement?: HTMLElement
  queries?: Q
  hydrate?: boolean
}

type DebugFn = (
  baseElement?: HTMLElement | HTMLElement[],
  maxLength?: number,
  options?: prettyFormat.OptionsReceived
) => void

type Result<Q extends Queries = typeof allQueries> = {
  container: HTMLElement
  baseElement: HTMLElement
  debug: DebugFn
  unmount: () => void
} & { [P in keyof Q]: BoundFunction<Q[P]> }

afterEach(() => cleanup())

const mountedContainers = new Set<Ref>()

function render<Q extends Queries = typeof allQueries>(ui: Ui, options: Options<Q>): Result<Q>
function render(ui: Ui, options?: Omit<Options, 'queries'>): Result

function render(ui: Ui, options: Options = {}): unknown {
  let { container, baseElement = container ?? document.body, queries, hydrate = false } = options

  container ??= baseElement.appendChild(document.createElement('div'))

  const unmount = (): void => {
    if (container?.parentElement === document.body) container.remove()
  }

  const dispose = hydrate ? solidHydrate(ui, container) : solidRender(ui, container)

  mountedContainers.add({ container, unmount, dispose })

  const debug: DebugFn = (el = baseElement, maxLength, options) =>
    Array.isArray(el)
      ? el.forEach((e) => console.log(prettyDOM(e, maxLength, options)))
      : console.log(prettyDOM(el, maxLength, options))

  return {
    container,
    baseElement,
    debug,
    unmount: dispose,
    ...getQueriesForElement(baseElement, queries),
  }
}

function cleanup(): void {
  mountedContainers.forEach(cleanupAtContainer)
  mountedContainers.clear()
}

function cleanupAtContainer(ref: Ref): void {
  ref.dispose()
  ref.unmount()
  mountedContainers.delete(ref)
}

export { render, cleanup }
