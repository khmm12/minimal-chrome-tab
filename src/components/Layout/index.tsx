import type { JSX } from 'solid-js'
import * as css from './styles'

interface LayoutProps {
  children?: [header: JSX.Element, content: JSX.Element, footer: JSX.Element]
}

export default function Layout(props: LayoutProps): JSX.Element {
  return <div class={css.container}>{props.children}</div>
}
