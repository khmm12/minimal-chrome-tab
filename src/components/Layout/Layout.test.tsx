import { render } from '@solidjs/testing-library'
import Layout from '.'

describe('Layout', () => {
  it('renders correctly', () => {
    const { container } = render(() => (
      <Layout>
        <div>Header</div>
        <div>Content</div>
        <div>Footer</div>
      </Layout>
    ))

    expect(container.firstChild).toMatchSnapshot()
  })
})
