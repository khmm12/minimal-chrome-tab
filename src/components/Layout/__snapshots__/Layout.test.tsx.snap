// Vitest Snapshot v1

exports[`Layout > renders correctly 1`] = `
<div
  class="container_c184xznc"
>
  <div>
    Header
  </div>
  <div>
    Content
  </div>
  <div>
    Footer
  </div>
</div>
`;
